src/aspire/abinitio/commonline_base.cu

```diff
@@ -0,0 +1,151 @@
+#include <stdint.h>
+#include <math.h>
+#include <cupy/complex.cuh>
+
+extern "C" __global__
+void build_clmatrix_kernel(
+    const int n,
+    const int m,
+    const int r,
+    const complex<double>* __restrict__ pf,
+    int16_t* const __restrict__ clmatrix,
+    const int n_shifts,
+    const complex<double>* const __restrict__ shift_phases)
+{
+  /* n n_img */
+  /* m angular componentns, n_theta//2 */
+  /* r radial componentns */
+  /* (n, m, r) = pf.shape in python (before transpose for CUDA kernel) */
+
+  /* thread index (2d), represents "i" and "j" indices */
+  const unsigned int i = blockDim.x * blockIdx.x + threadIdx.x;
+  const unsigned int j = blockDim.y * blockIdx.y + threadIdx.y;
+
+  /* no-op when out of bounds */
+  if(i >= n) return;
+  if(j >= n) return;
+  /* no-op lower triangle */
+  if(j <= i) return;
+
+  int k;
+  int s;
+  int cl1, cl2;
+  int best_cl1, best_cl2;
+  double dist, best_cl_dist;
```

Comment:
I'm guessing this is inherited from Python/Matlab, but these aren't distances, are they? More like correlations?

Reply:
Name is from MATLAB but the code isn't. They are dot products, so yes, they form the set of cross-correlations. Changed.

Reply:
Thanks!